src/Adapter/Order/Refund/OrderProductRemover.php

```diff
@@ -0,0 +1,265 @@
+        float $shippingDiffTaxIncl,
+        float $shippingDiffTaxExcl
+    ) {
+        $order->total_shipping -= $shippingDiffTaxIncl;
+        $order->total_shipping_tax_excl -= $shippingDiffTaxExcl;
+        $order->total_shipping_tax_incl -= $shippingDiffTaxIncl;
+        $order->total_products -= $shippingDiffTaxExcl;
+        $order->total_products_wt -= $productPriceTaxIncl;
+        $order->total_paid -= $productPriceTaxIncl + $shippingDiffTaxIncl;
+        $order->total_paid_tax_incl -= $productPriceTaxIncl + $shippingDiffTaxIncl;
+        $order->total_paid_tax_excl -= $productPriceTaxExcl + $shippingDiffTaxExcl;
+        $order->total_paid_real -= $productPriceTaxIncl + $shippingDiffTaxIncl;
+
+        $fields = [
+            'total_shipping',
+            'total_shipping_tax_excl',
+            'total_shipping_tax_incl',
+            'total_products',
+            'total_products_wt',
+            'total_paid',
+            'total_paid_tax_incl',
+            'total_paid_tax_excl',
+            'total_paid_real',
+        ];
+
+        /* Prevent from floating precision issues */
+        foreach ($fields as $field) {
+            if ($order->{$field} < 0) {
+                $order->{$field} = 0;
+            }
+        }
+
+        /* Prevent from floating precision issues */
+        foreach ($fields as $field) {
+            $order->{$field} = number_format(
+                $order->{$field},
+                Context::getContext()->getComputingPrecision(),
+                '.',
+                ''
+            );
+        }
+    }
+
+    /**
+     * @param Order $order
+     * @param OrderDetail $orderDetail
+     * @param int $quantity
+     * @param float $productPriceTaxIncl
+     * @param float $productPriceTaxExcl
+     * @param float $shippingDiffTaxIncl
+     * @param float $shippingDiffTaxExcl
+     *
+     * @throws DeleteProductFromOrderException
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function updateOrderDetail(
+        Order $order,
+        OrderDetail $orderDetail,
+        int $quantity,
+        float $productPriceTaxIncl,
+        float $productPriceTaxExcl,
+        float $shippingDiffTaxIncl,
+        float $shippingDiffTaxExcl
+    ) {
+        $orderDetail->product_quantity -= (int) $quantity;
+        if ($orderDetail->product_quantity == 0) {
+            if (!$orderDetail->delete()) {
+                throw new DeleteProductFromOrderException('Could not delete order detail');
+            }
+            if (count($order->getProductsDetail()) == 0) {
+                $history = new OrderHistory();
+                $history->id_order = (int) $order->id;
+                $history->changeIdOrderState(Configuration::get('PS_OS_CANCELED'), $order);
+                if (!$history->addWithemail()) {
+                    // email failure must not block order update process
+                    $this->logger->warning(
```

Comment:
This logging is nice, but do you think there would be a way to display it to the user?
The delete action would work fine (displaying a success block), but it would also display a warning block regarding the mail sending failure

Reply:
As seen with @eternoendless and the product team , the rule we stood by is:

If mail failure is blocking the feature: error message
If not:  logging message (as a warning)

Being able to send emails is not mandatory for a shop.